### PR TITLE
Further linux compatibility

### PR DESCRIPTION
**Pull request by 20PercentRendered**
This PR allows you to fully build Nitrox on linux, provided you have dotnet, net6, mono and mono-msbuild installed. 
The built binaries work flawlessly on linux, and should work on Windows too.

This PR is ready to be reviewed and merged.

I've decided to make Nitrox.CLI as an [external git repo](https://github.com/20PercentRendered/Nitrox.CLI) I can clone into the Nitrox source tree if I want to do prototyping.

### CODE REVIEW

**Comment by Measurity** [comment]
Thanks for helping out.

> Should I also make a Nitrox.CLI for launching the game and server without a GUI launcher (potentially useful for steamdeck)?
We can use the new Nitrox.Launcher project and pass arguments to it without starting up the UI (for performance reasons). But we're not done with reimplementing these features yet (for both windows and linux) anyway.

**Comment by 20PercentRendered** [comment]
A CLI would be useful for prototyping the rest of Linux support to a working state before the new launcher is ready. Like proton prefix management

**Comment by Measurity** [comment]
The new launcher is easy enough to prototype because much of the functionality we need is already done in the old launcher.

I could see a CLI being useful as an alternative to the launcher for letting Nitrox install/repair itself or returning the path to the game via commands. Especially for server owners. But I'm not convinced yet why this can't be done with the existing projects (server is already a console project). So I'd like to delay the CLI project until after the launcher is done and we have a better idea what to add to the CLI.

**Comment by 20PercentRendered** [comment]
Currently, the server runs on mono and libgdiplus for the RandomStartGenerator, however in .NET Core, System.Drawing is not supported. I'll make a separate MR for porting the servers to net6. 
But right now the bigger issue is dotnet run not starting the server inside the build folder. That will be the last thing to fix in this MR.

**Comment by 20PercentRendered** [comment]
I haven't tested building my branch using VS 2022 or JB Rider, but these changes are pretty trivial so it should work. Otherwise, this is perfectly ready for review.

**Comment by 20PercentRendered** [comment]
This PR is ready to be reviewed and merged.

**Comment by 20PercentRendered** [comment]
Just tested, this PR works ingame. 

Note for implementing the Nitrox.Launcher project: On linux, you should symlink the proton prefix's `AppData/Roaming/Nitrox` folder to some other folder on the native side, such as `.local/share/Nitrox`. Wine/Proton doesn't seem to like creating an actual folder there and using it. 
If this isn't done, the servers list will not load. 

Also, running steam with `-applaunch 264710 -nitrox /path/to/nitrox/build` will run the game with Nitrox enabled, if Nitrox was already modified into `Assembly-CSharp`.

File: NitroxModel/Helper/NitroxUser.cs
Comment (Measurity) [inline NitroxModel/Helper/NitroxUser.cs:87]: Since `RegistryEx` is our code, I would prefer we refactor that to make it compatible as a key/value store for Linux (using something on linux that is conventional for key/value config store). Renaming it to something more generic for both OSes. This would reduce the amount of code lines for Linux checking elsewhere.

File: NitroxModel/Logger/Log.cs
Comment (Measurity) [inline NitroxModel/Logger/Log.cs:83]: 1. The indentation does not conform the rest of the code.
2. The logger shouldn't be added if it needs to be skipped, as an optimization.

File: NitroxModel/Helper/NitroxUser.cs
Comment (Measurity) [inline NitroxModel/Helper/NitroxUser.cs:27]: Does not conform code style for braces on separate lines and variable naming (no `_`). Or could be rewritten as:
```suggestion
        private static LinuxConfig linuxConfig;
        public static LinuxConfig LinuxConfig => linuxConfig ??= LinuxConfig.Load();
```

File: NitroxModel/Helper/LinuxConfig.cs
Comment (Measurity) [inline NitroxModel/Helper/LinuxConfig.cs:24]: Not needed if refactoring `RegisteryEx` to work on Linux.

File: Nitrox.Launcher/Nitrox.Launcher.csproj
Comment (Measurity) [inline Nitrox.Launcher/Nitrox.Launcher.csproj:35]: Not a fan of changing the version to a variable here. We use IDEs that interface with nuget to handle upgrade packages. Any change to the generated `<PackageReference>` will be reverted (leaving the `AvaloniaVersion` variable useless)

File: Nitrox.Launcher/Nitrox.Launcher.csproj
Comment (Measurity) [inline Nitrox.Launcher/Nitrox.Launcher.csproj:4]: `TargetFramework` is very project-specific and only used in Nitrox.Launcher at the moment. Would prefer no additional code unless there's at least 2 or more instances where it is used.

File: Directory.Build.props
Comment (Measurity) [inline Directory.Build.props:8]: `NitroxNetCoreVersion` can be removed. (Minor but `.NET` is the new name for `.NET core`).

File: .vscode/launch.json
Comment (Measurity) [inline .vscode/launch.json:12]: Since nobody on our team uses VSCode at the moment, it would be good to document where this template was copied from. Maybe even add a page to the Nitrox wiki later.

**Review by Measurity — CHANGES_REQUESTED** [review]
The idea of the `linux` branch was to first work on the launcher UI and features, and then migrate over to it while the rest of the project makes the transition to .netstandard 2.0.

I'm hesitant to merge changes to other projects (not `Nitrox.Launcher`) as it can be a while for the `linux` branch to get merged (still need to do a bunch of work on the UI). 

In addition, I prefer using the same code components (like `RegistryEx`) on all OSes instead of if-checking specific OS code around using these APIs. The latter can lead to spaghetti code.

File: Nitrox.BuildTool/Program.cs
Comment (20PercentRendered) [reply Nitrox.BuildTool/Program.cs:92]: NitroxModel calls Log functions when finding your install. While Log may not be directly used by Nitrox.BuildTool, it needs to be set up in BuildTool so that logging in NitroxModel works inside BuildTool.

File: Nitrox.Launcher/Nitrox.Launcher.csproj
Comment (20PercentRendered) [reply Nitrox.Launcher/Nitrox.Launcher.csproj:86]: The only way to fix these would be to use net472 for the launcher, or port most Nitrox.* projects to net6, which isn't possible due to these being required in Unity-side code. We could also decouple building Nitrox's clientside libs and the libs required for the launcher, but this would result in two DLLs with identical functionality.

File: NitroxModel/Helper/NitroxUser.cs
Comment (20PercentRendered) [reply NitroxModel/Helper/NitroxUser.cs:87]: Sounds like a good idea. Will make changes.

File: Nitrox.Launcher/Nitrox.Launcher.csproj
Comment (Measurity) [reply Nitrox.Launcher/Nitrox.Launcher.csproj:86]: Separating game and server DLLs is indeed the way to go. The game has some support for netstandard 2.0 with latest subnautica update (living large). But not all code is compatible (Serilog causes a `TypeLoadException` from my testing).

File: Nitrox.Launcher/Nitrox.Launcher.csproj
Comment (20PercentRendered) [reply Nitrox.Launcher/Nitrox.Launcher.csproj:4]: I intend to port the server and buildtool to net6 as well. Would this variable still be worth keeping around?

File: NitroxModel/Helper/NitroxUser.cs
Comment (20PercentRendered) [reply NitroxModel/Helper/NitroxUser.cs:87]: Should I keep RegistryEx around, and create a new KeyValueConfig class, that uses RegistryEx on Windows and .config on linux?

File: Nitrox.Launcher/Nitrox.Launcher.csproj
Comment (Measurity) [reply Nitrox.Launcher/Nitrox.Launcher.csproj:86]: > The servers could be built separately and then the launcher would just run the exe.
This is how we want to do it. Start the exe with some arguments/environment-variables, and then never interface with server again other than packet serialization.

I'm in the works with splitting client and server up and be build separately. But the serializer (BinaryPack) we use for packets doesn't work with netstandard 2.0 which is currently blocking me.

If you keep the .csproj changes for later I would appreciate it.

File: NitroxModel/Helper/NitroxUser.cs
Comment (Measurity) [reply NitroxModel/Helper/NitroxUser.cs:87]: Yes that's also good with me. Up to you. Either rename RegistryEx or wrap it in another API.

File: NitroxModel/Logger/Log.cs
Comment (20PercentRendered) [reply NitroxModel/Logger/Log.cs:83]: About point 2: The code above it does the exact same `return;`-style way of toggling it. Will fix point 1 though.

Comment (Measurity) [reply NitroxModel/Logger/Log.cs:83]: True, should also be changed. Here's a pastebin with new extension methods to optimize this code and reduce `Setup` method code size:
https://pastebin.com/MhGHQtcb

File: Nitrox.Launcher/Nitrox.Launcher.csproj
Comment (20PercentRendered) [reply Nitrox.Launcher/Nitrox.Launcher.csproj:86]: Will hold off the csproj changes.

File: Nitrox.Launcher/Nitrox.Launcher.csproj
Comment (20PercentRendered) [reply Nitrox.Launcher/Nitrox.Launcher.csproj]: Variable removed in newest commit. In most cases however, you want to avoid multiple versions of .NET Core, as it causes more burden to the users who need to have multiple runtimes installed.

File: .vscode/launch.json
Comment (Measurity) [reply .vscode/launch.json:12]: The team has decided that we prefer not include .vscode files until there are more people using it. This because it will likely break and we don't have any maintainers at this time to update it. You mind if you add the .vscode folder to .gitignore for now until it gets more use?

Comment (20PercentRendered) [reply .vscode/launch.json:12]: I don't mind. However, there's very little that can actually break in the .vscode files, since they essentially do two things:
1. Run `dotnet build` on the project 
2. Instruct Microsoft's .NET debugger where the dll is
The .NET debugger then launches the dll with the debugger attached. Unless Microsoft changes the debuggers name from "coreclr" to something else, or changes how the JSON works, nothing can break. 
The JSON files are generated from the C# extension's template, which I've just copy-and-pasted for all the Nitrox projects. 
To try it out, delete the .vscode folder and from the sidebar click the "Debug" button and it will prompt you to create a launch configuration, which it will then auto fill.

Comment (20PercentRendered) [reply .vscode/launch.json:12]: And also, vscode is basically the only viable C# editor, since VS doesn't exist, and JB Rider costs money. Not to mention most linux users prefer open-source free software.

Comment (Measurity) [reply .vscode/launch.json:12]: I agree with your reasoning. We did a rethink, but still slightly prefer more to not have it. This could easily change in the future, or there could be alternatives (reusing files from other IDEs like .editorconfig).

Comment (20PercentRendered) [reply .vscode/launch.json:12]: VSCode's C# extension can read .editorconfig, but it needs a setting to enable it. This setting should be stored at workspace-level .vscode. 

Note that VSCode by default also doesn't enable analyzers, and needs a setting to enable it. I've commited some changes that allow for these to happen.

If these settings don't exist, anyone using VSCode will commit code that isn't formatted according to `CONTRIBUTING.md`, and doesn't use analyzers, leaving the user guessing at what is correct formatting.

File: NitroxModel/Helper/NitroxUser.cs
Comment (20PercentRendered) [reply NitroxModel/Helper/NitroxUser.cs:87]: Should I use System.Text.Json for the .config file? Attempting to use NitroxModel (net472 project) from Nitrox.Launcher (net6.0 project) for testing the .config file approach causes a runtime error: `Could not load file or assembly 'System.Text.Json, Version=7.0.0.1, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. The located assembly's manifest definition does not match the assembly reference. (0x80131040)`. Changing NitroxModel's `TargetFramework` to `<TargetFrameworks>net472;net6.0</TargetFrameworks>` fixes this. This change doesn't seem to affect anything else negatively. Using an XMLDocument is not possible, as dictionaries can't be serialized, and adding Newtonsoft.Json could cause headaches, as this is used by the game.

Comment (20PercentRendered) [reply NitroxModel/Helper/NitroxUser.cs:87]: Nevermind. Fixed by downgrading System.Text.Json from 7.0.1 to 7.0.0.

File: NitroxModel/Logger/Log.cs
Comment (20PercentRendered) [reply NitroxModel/Logger/Log.cs]: Fixed in newest commits.

File: NitroxModel/Helper/NitroxUser.cs
Comment (20PercentRendered) [reply NitroxModel/Helper/NitroxUser.cs]: Addressed in newest commits.

File: NitroxModel/Helper/LinuxConfig.cs
Comment (20PercentRendered) [reply NitroxModel/Helper/LinuxConfig.cs]: Addressed in newest commits.

File: NitroxModel/Helper/NitroxUser.cs
Comment (20PercentRendered) [reply NitroxModel/Helper/NitroxUser.cs]: Addressed in newest commits.

File: .vscode/launch.json
Comment (Measurity) [reply .vscode/launch.json]: Let's keep it out of this PR for now. I'll add you to the team and you can discuss it with the team yourself. Right now the consensus is no.

Comment (20PercentRendered) [reply .vscode/launch.json]: Seems good. Only thing left to review is the new key-value store.

**Review by Measurity — APPROVED** [review]

**Comment by 20PercentRendered** [comment]
Changed KeyValueStore slightly, needs rereview.

File: NitroxModel/Helper/KeyValueStore.cs
Comment (Measurity) [inline NitroxModel/Helper/KeyValueStore.cs:56]: Accessing local state (`instance` variable) in `GetKeyValueStoreForPlatform` isn't necessary. 
`public static IKeyValueStore Instance { get; } = GetKeyValueStoreForPlatform();` means that the instance will be set when a new instance of `KeyValueStore` is created (constructor is called).

Comment (Measurity) [reply NitroxModel/Helper/KeyValueStore.cs:56]: (unfortunately, mono runtime used by Unity does not support C# 10 `static abstract members` on interfaces which would have cleaned up this code)

Comment (Measurity) [reply NitroxModel/Helper/KeyValueStore.cs]: I committed the changes

Comment (20PercentRendered) [reply NitroxModel/Helper/KeyValueStore.cs]: Yep, it was a different bug.

**Review by Measurity — APPROVED** [review]

**Comment by 20PercentRendered** [comment]
This PR is now ready. I could still look at ProcessEx and some other APIs, but I'll hold off those for some other time. Maybe once the new launcher is ready enough to launch the game on Windows. If you need linux code or testing, feel free to tag me.

**Pull request merged**
